pkgs/by-name/ta/tabby/package.nix

```diff
@@ -1,36 +1,141 @@
-{ lib
+{ config
+, lib
+, rustPlatform
 , fetchFromGitHub
-, gcc12
-, cmake
+, stdenv
+
 , git
 , openssl
 , pkg-config
 , protobuf
-, rustPlatform
-, addOpenGLRunpath
-, cudatoolkit
-, nvidia ? true
+
+, llama-cpp
+
+, cudaSupport ? config.cudaSupport
+, cudaPackages ? { }
+
+, rocmSupport ? config.rocmSupport
+
+, darwin
+, metalSupport ? stdenv.isDarwin && stdenv.isAarch64
+
+  # one of [ null "cpu" "rocm" "cuda" "metal" ];
+, acceleration ? null
 }:
 
-rustPlatform.buildRustPackage rec {
-  version = "0.7.0";
+let
+  inherit (lib) optional optionals flatten;
+  # References:
+  # https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/ll/llama-cpp/package.nix
+  # https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/misc/ollama/default.nix
+
   pname = "tabby";
+  version = "0.8.3";
+
+
+  availableAccelerations = flatten [
+    (optional cudaSupport "cuda")
+    (optional rocmSupport "rocm")
+    (optional metalSupport "metal")
+  ];
+
+  warnIfMultipleAccelerationMethods = configured: (let
+    len = builtins.length configured;
+    result = if len == 0 then "cpu" else (builtins.head configured);
+  in
+    # lib.warnIf (len > 1) "multiple acceleration methods available: ${toString availableAccelerations}" "cpu"
+    lib.warnIf (len > 1) ''
+      building tabby with multiple acceleration methods is not supported;
+      falling back to `${result}`
+
+      To configure tabby with a specific acceleration method either
+
+      1. configure a package override
+      2. set `services.tabby.acceleration`
+
+      For example:
+        services.tabby.acceleration = "cuda";
+      Or:
+        services.tabby.acceleration = "rocm";
```

Comment:
It feels backward to mention NixOS module configuration in the warning of a specific package. Perhaps this validation deserves to be in the module instead. A shorter, less-verbose warning could be here instead.

Reply:
Yeah that makes sense, I'll move this around

Reply:
Looks better. Thanks. It looks like you have a trailing whitespace somewhere, causing the CI to fail.